Source/WebCore/bindings/js/JSWindowProxy.cpp

```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
     return &static_cast<JSVMClientData*>(vm.clientData)->windowProxySpace();
 }
 
+struct FrameInfo {
+    Frame& frame;
+    Frame& mainFrame;
+};
+
+static std::optional<FrameInfo> getFrameInfo(JSGlobalObject* globalObject)
+{
+    auto* domGlobalObject = jsDynamicCast<JSDOMGlobalObject*>(globalObject);
+    if (!domGlobalObject)
+        return std::nullopt;
+
+    auto* document = dynamicDowncast<Document>(domGlobalObject->scriptExecutionContext());
+    if (!document)
+        return std::nullopt;
+
+    auto* frame = document->frame();
+    if (!frame)
+        return std::nullopt;
+
+    return FrameInfo { *frame, frame->mainFrame() };
```

Comment:
Does the caller gracefully handle if the frame goes away?

Reply:
I wouldn't think that is possible since I'm not calling back out to JS or something that could cause a deallocation from underneath me. But I don't really know what else can cause a Frame to disappear beneath me and I've been told we're pretty much banned from using raw refs and pointers except for trivial use cases, so I'll switch these to use Refs instead.